### PR TITLE
Refactor object-detection models

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?`

Improves code readability for object-detection models, should be merged after 
 - #36460 
 
**Note:**

> RT-DETR and RT-DETR-v2 are refactored, for other models just Freezed BatchNorm2d module updated.

**RT-DETR refactoring:** 

  - Make code closer to `transformers` standards (no one-letter vars, common var names, ...)
  - Remove some unused outputs (seems not necessary, but add overhead on implementation and docs)
  - Clean up signatures from unused arguments (where it's possible)
  - Better comments and shape comments
  - Add default docstring for object-detection
  - Unify RT-DETR and RT-DETRv2 implementations (simpler modular file for v2)

**In addition:**
 
 - Fix CoreML export for RT-DETR (avoid 6D tensors in deformable_attention)
 
Mostly, the changes are backward compatible, but some **internal** modules' signatures have changed. I suspect RT-DETR **internal** modules are used on their own somewhere outside the `transformers` repository. A quick search on GitHub shows only transformers forks, and I have never seen custom code samples on HF Hub for object detection. Therefore, I think we can change them to establish better standards for upcoming models such as D-Fine (a state-of-the-art real-time object detector) and RT-DETRv3, both of which are based on RT-DETR.

### Test changes

I tried to keep tests untouched, but there are a few modifications:
 - the number of outputs in ModelOuput is reduced

## Who can review?

cc @SangbumChoi @jadechoghari as model contributors, in case you want to have a look. Please let me know if anything is missing / incorrectly re-implemented

### CODE REVIEW

**Review by SangbumChoi — COMMENTED** [review]
Good :) (In general I think the PR is in still draft.)

File: src/transformers/models/conditional_detr/modeling_conditional_detr.py [hunk 1]
```diff
@@ -363,8 +362,8 @@ def __init__(self, config):
             backbone = load_backbone(config)
 
         # replace batch norm by frozen batch norm
-        with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/conditional_detr/modeling_conditional_detr.py:366]: No, we are just copying tensor data, so no need for no-grad. I tested with no-grad removed, replace_batch_norm works fine

File: src/transformers/models/rt_detr/modeling_rt_detr.py [hunk 2]
```diff
@@ -326,141 +224,411 @@ class RTDetrObjectDetectionOutput(ModelOutput):
     pred_boxes: torch.FloatTensor = None
     auxiliary_outputs: Optional[List[Dict]] = None
     last_hidden_state: torch.FloatTensor = None
-    intermediate_hidden_states: torch.FloatTensor = None
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/rt_detr/modeling_rt_detr.py:329]: Thanks, I will double-check, but it seems redundant. As far as I recall it is similar to encoder hidden states (shifted by one, because taken after the layer applied)

File: src/transformers/models/rt_detr/modeling_rt_detr.py [hunk 3]
```diff
@@ -708,194 +795,36 @@ def __init__(self, config: RTDetrConfig):
...
-        reference_points=None,
-        spatial_shapes=None,
-        spatial_shapes_list=None,
-        level_start_index=None,
-        output_attentions: bool = False,
-    ):
-        # add position embeddings to the hidden states before projecting to queries and keys
-        if position_embeddings is not None:
-            hidden_states = self.with_pos_embed(hidden_states, position_embeddings)
+        params = {"kernel_size": 1, "stride": 1, "activation": config.activation_function}
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/rt_detr/modeling_rt_detr.py:799]: I agree it might not be the best approach, I did it just for code readability:
 - No line breaks for module definitions (in my opinion, it's a bit easier to read)
 - It also shows that parameters are identical across modules

However, it's very subjective.
